src/go/k8s/internal/controller/redpanda/index.go

```diff
@@ -0,0 +1,59 @@
+package redpanda
+
+import (
+	"context"
+
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	userClusterIndex = "__user_referencing_cluster"
+)
+
+func userCluster(user *redpandav1alpha2.User) types.NamespacedName {
+	return types.NamespacedName{Namespace: user.Namespace, Name: user.Spec.ClusterSource.ClusterRef.Name}
+}
+
+func registerUserClusterIndex(ctx context.Context, mgr ctrl.Manager) error {
+	return mgr.GetFieldIndexer().IndexField(ctx, &redpandav1alpha2.User{}, userClusterIndex, indexUserCluster)
+}
```

Comment:
the general pattern I've always followed is to put both the `IndexField` call and the `List` calls leveraging the index into helpers within a a file that contains all of the indexing logic. Part of this is due to the need to pass in:

```
FieldSelector: fields.OneTermEqualSelector(INDEX_NAME, VALUE)
```

on calls that leverage the index, and I'd rather not leak all of the funny index name constants/invocations over the rest of the code.